File: 042 Trapping Rain Water.c
#include <stdio.h>
#include <stdlib.h>
/*
Trapping Rain Water
https://oj.leetcode.com/problems/trapping-rain-water/
http://yucoding.blogspot.com/2013/05/leetcode-question-111-trapping-rain.html
*/
#define MAX(a,b) (a)>(b)?(a):(b)
#define MIN(a,b) (a)<(b)?(a):(b)

int trap(int A[], int n)
{
    int *l,*r, t, i,water;
    if (n<2) return 0; 
     
    l = (int*)malloc(sizeof(int)*n);
    r = (int*)malloc(sizeof(int)*n);
    
    l[0]=0;
    for(i=1;i < n;i++){
        l[i] = MAX(A[i-1],l[i-1]);
    }
    
    r[n-1] = 0;
    for(i = n-2;i > -1;i--){
        r[i] = MAX(A[i+1],r[i+1]);
    }
    
    water = 0;
    for(i=1;i < n-1;i++){
        t = MIN(l[i],r[i]);
        t -= A[i];
        water += MAX(t,0);
    }  
    free(l);
    free(r);
    return water;
}
/*
Tester
*/
int main()
{
	int arr[] = {0,1,0,2,1,0,1,3,2,1,2,1};
    int arr_size = sizeof(arr)/sizeof(arr[0]);
    printf("%d\n", trap(arr, arr_size));
    return 0;
}

File: 042 Trapping Rain Water.py
'''
42. Trapping Rain Water
https://leetcode.com/problems/trapping-rain-water/

Given n non-negative integers representing an elevation map where the width of each bar is 1,
compute how much water it can trap after raining.

Example 1:

Input: height = [0,1,0,2,1,0,1,3,2,1,2,1]
Output: 6
Explanation: The above elevation map (black section) is represented by array [0,1,0,2,1,0,1,3,2,1,2,1].
In this case, 6 units of rain water (blue section) are being trapped.

Example 2:

Input: height = [4,2,0,3,2,5]
Output: 9

Constraints:
    n == height.length
    1 <= n <= 2 * 104
    0 <= height[i] <= 105
'''
class Solution:
    '''
    Use Prefix Sum. This approach needs O(n) space
    '''
    def trap(self, height: List[int]) -> int:
        '''
        The maximum water on index i is the minimum of (highest height on its left
        and highest height on its right) - height on i
        '''
        leftHighs = [0] * len(height) # leftHighs[i] is the highest in [:i]
        for i in range(1, len(height)):
            leftHighs[i] = max(leftHighs[i-1], height[i-1])
            
        rightHighs = [0] * len(height) # rightHighs[i] is the highest in [i+1:]
        for i in range(len(height)-2, -1, -1):
            rightHighs[i] = max(rightHighs[i+1], height[i+1])
            
        result = 0
        for i in range(len(height)):
            result += max(0, min(leftHighs[i], rightHighs[i]) - height[i])
        return result

class Solution2:
    '''
    Two pointers approach, O(1) space
    '''
    def trap(self, heights: List[int]) -> int:
        left = 0
        right = len(heights) - 1
        leftHighest, rightHighest = 0, 0
        water = 0
        while left < right:
            if heights[left] < heights[right]:
                # Index left can possibly hold water as there is a higher right
                if heights[left] > leftHighest:
                    # Index left cannot hold water as it does not have a higher left
                    leftHighest = heights[left] 
                else:
                    # It must be leftHighest < rightHighest, as it only moves left when eights[left] < heights[right] 
                    water += leftHighest - heights[left]
                left += 1
            else: # heights[left] >= heights[right]
                if heights[right] > rightHighest:
                    rightHighest = heights[right]
                else:
                    water += rightHighest - heights[right]
                right -= 1

        # Eventually, left and right will meet in leftHighest or rightHighest
        return water

